File: internal/tools/cli_test.go
package tools

import (
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/iamangus/code-mcp/internal/worktree"
)

// TestExecuteTerminalCommand_EchoStdout verifies stdout capture.
func TestExecuteTerminalCommand_EchoStdout(t *testing.T) {
	dir := t.TempDir()
	stdout, stderr, code, timedOut, err := ExecuteTerminalCommand(dir, "echo hello", 5*time.Second)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if timedOut {
		t.Fatal("should not have timed out")
	}
	if code != 0 {
		t.Errorf("expected exit code 0, got %d", code)
	}
	if !strings.Contains(stdout, "hello") {
		t.Errorf("expected 'hello' in stdout, got %q", stdout)
	}
	_ = stderr
}

// TestExecuteTerminalCommand_NonZeroExit verifies non-zero exit code capture.
func TestExecuteTerminalCommand_NonZeroExit(t *testing.T) {
	dir := t.TempDir()
	_, _, code, timedOut, err := ExecuteTerminalCommand(dir, "exit 42", 5*time.Second)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if timedOut {
		t.Fatal("should not have timed out")
	}
	if code != 42 {
		t.Errorf("expected exit code 42, got %d", code)
	}
}

// TestExecuteTerminalCommand_StderrCapture verifies stderr capture.
func TestExecuteTerminalCommand_StderrCapture(t *testing.T) {
	dir := t.TempDir()
	_, stderr, _, _, err := ExecuteTerminalCommand(dir, "echo error_msg >&2", 5*time.Second)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !strings.Contains(stderr, "error_msg") {
		t.Errorf("expected 'error_msg' in stderr, got %q", stderr)
	}
}

// TestExecuteTerminalCommand_Timeout verifies that the timeout flag is set.
func TestExecuteTerminalCommand_Timeout(t *testing.T) {
	dir := t.TempDir()
	_, _, _, timedOut, _ := ExecuteTerminalCommand(dir, "sleep 10", 1*time.Millisecond)
	if !timedOut {
		t.Error("expected timeout flag to be true")
	}
}

// TestExecuteTerminalCommand_InvalidDir verifies ToolError for bad worktree.
func TestExecuteTerminalCommand_InvalidDir(t *testing.T) {
	_, _, _, _, err := ExecuteTerminalCommand("/nonexistent/dir/xyz", "echo hi", 5*time.Second)
	if err == nil {
		t.Fatal("expected error for invalid dir, got nil")
	}
	if _, ok := err.(*worktree.ToolError); !ok {
		t.Fatalf("expected ToolError, got %T", err)
	}
}

// TestGetGitDiff_InGitRepo verifies git diff output in a temporary git repository.
func TestGetGitDiff_InGitRepo(t *testing.T) {
	if _, err := exec.LookPath("git"); err != nil {
		t.Skip("git not available")
	}

	dir := t.TempDir()

	run := func(args ...string) {
		t.Helper()
		cmd := exec.Command("git", args...)
		cmd.Dir = dir
		cmd.Env = append(os.Environ(),
			"GIT_AUTHOR_NAME=Test",
			"GIT_AUTHOR_EMAIL=test@example.com",
			"GIT_COMMITTER_NAME=Test",
			"GIT_COMMITTER_EMAIL=test@example.com",
		)
		if out, err := cmd.CombinedOutput(); err != nil {
			t.Fatalf("git %v failed: %v\n%s", args, err, out)
		}
	}

	run("init")
	run("config", "user.email", "test@example.com")
	run("config", "user.name", "Test")

	// Create and commit initial file
	filePath := filepath.Join(dir, "hello.txt")
	os.WriteFile(filePath, []byte("initial\n"), 0644)
	run("add", "hello.txt")
	run("commit", "-m", "initial commit")

	// Modify the file (unstaged change)
	os.WriteFile(filePath, []byte("modified\n"), 0644)

	out, err := GetGitDiff(dir)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	// Should show either diff or status
	if out == "" {
		t.Error("expected non-empty output from GetGitDiff")
	}
}

// TestGetGitDiff_InvalidDir verifies ToolError for bad worktree.
func TestGetGitDiff_InvalidDir(t *testing.T) {
	_, err := GetGitDiff("/nonexistent/dir/xyz")
	if err == nil {
		t.Fatal("expected error for invalid dir, got nil")
	}
	if _, ok := err.(*worktree.ToolError); !ok {
		t.Fatalf("expected ToolError, got %T", err)
	}
}

File: src/__tests__/cli.test.ts
import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import { WorktreeJail } from '../worktree.js';
import { executeTerminalCommand, getGitDiff } from '../tools/cli.js';
import fs from 'fs';
import os from 'os';
import path from 'path';
import { execSync } from 'child_process';

let tmpDir: string;
let jail: WorktreeJail;

beforeAll(() => {
  tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'cli-test-'));
  jail = new WorktreeJail(tmpDir);
});

afterAll(() => {
  fs.rmSync(tmpDir, { recursive: true, force: true });
});

describe('execute_terminal_command', () => {
  it('runs echo and captures stdout', async () => {
    const result = await executeTerminalCommand(jail, 'echo "hello world"');
    expect(result.stdout.trim()).toBe('hello world');
    expect(result.exitCode).toBe(0);
    expect(result.timedOut).toBe(false);
  });

  it('captures exit code', async () => {
    const result = await executeTerminalCommand(jail, 'exit 42');
    expect(result.exitCode).toBe(42);
  });

  it('captures stderr', async () => {
    const result = await executeTerminalCommand(jail, 'echo "error message" >&2');
    expect(result.stderr.trim()).toBe('error message');
  });

  it('times out with very short timeout', async () => {
    const result = await executeTerminalCommand(jail, 'sleep 10', 100);
    expect(result.timedOut).toBe(true);
    expect(result.exitCode).toBe(-1);
  });
});

describe('get_git_diff', () => {
  it('returns diff string in a git repo', async () => {
    execSync('git init', { cwd: tmpDir });
    execSync('git config user.email "test@test.com"', { cwd: tmpDir });
    execSync('git config user.name "Test"', { cwd: tmpDir });
    fs.writeFileSync(path.join(tmpDir, 'test.txt'), 'initial content');
    execSync('git add .', { cwd: tmpDir });
    execSync('git commit -m "initial"', { cwd: tmpDir });
    fs.writeFileSync(path.join(tmpDir, 'test.txt'), 'modified content');

    const diff = await getGitDiff(jail);
    expect(typeof diff).toBe('string');
    expect(diff).toContain('modified content');
  });
});
